### DISCUSSION TITLE
Question: What are the plans for the future of testing Windows apps?

### DISCUSSION THREAD

**Discussion by mrlacey** [microsoft/WindowsAppSDK#382] [Q&A]
Testing of UWP apps is an area that has been overlooked in the past. Tooling support for testing is lacking in many areas and little attention has been paid.

- Are plans being made to ensure that all apps built using Project Reunion are fully testable?
- Will all APIs be mockable?
- Will [MSTest Fakes](https://docs.microsoft.com/en-us/visualstudio/test/isolating-code-under-test-with-microsoft-fakes?view=vs-2019) support be added for all APIs?
- Will new project and item templates include options for including tests?
- Will work be done to ensure that work done with testing tools supports Windows apps in future?
- What else is being considered for software testing as part of this project?

I'm aware of #227 but this question is aimed at the broader situation, not just UI testing.

X-Ref: https://github.com/microsoft/microsoft-ui-xaml/issues/3584

**Comment by stevewri** [comment]
Really good questions, and this isn't an area we've thought deeply about yet. Let's begin by seeing if we can get some more discussion on this issue (and then turn it into one).

**Comment by rgwood** [comment]
Thanks for raising this @mrlacey, good questions. Personally I have often avoided writing tests for UWP code because:

1. Tests are slow to launch (likely related to #55)
1. My TDD tools (NCrunch in particular) do not support running UWP projects

**Comment by mrlacey** [comment]
@rgwood These are great examples of why this needs to be addressed.
